js/js.tests/test/org/jetbrains/kotlin/js/test/InteropJsTestChecker.kt

```diff
@@ -5,65 +5,61 @@
 import org.junit.Assert
-import javax.script.Invocable
-import javax.script.ScriptEngine
 
-fun createScriptEngine(): ScriptEngine =
-    // TODO use "-strict"
-    NashornScriptEngineFactory().getScriptEngine("--language=es5", "--no-java", "--no-syntax-extensions")
+private val USE_J2V8_INTEROP_FOR_JS_TESTS = java.lang.Boolean.getBoolean("org.jetbrains.kotlin.use.j2v8.interop.for.js.tests")
 
-fun ScriptEngine.overrideAsserter() {
-    eval("this['kotlin-test'].kotlin.test.overrideAsserter_wbnzx$(this['kotlin-test'].kotlin.test.DefaultAsserter);")
+fun createScriptEngine(): InteropEngine {
+    return if (USE_J2V8_INTEROP_FOR_JS_TESTS) {
+        InteropV8()
+    } else {
+        InteropNashorn()
+    }
+}
+
+fun InteropEngine.overrideAsserter() {
+    evalVoid("this['kotlin-test'].kotlin.test.overrideAsserter_wbnzx$(this['kotlin-test'].kotlin.test.DefaultAsserter);")
 }
 
-fun ScriptEngine.runTestFunction(
+fun InteropEngine.runTestFunction(
     testModuleName: String?,
     testPackageName: String?,
     testFunctionName: String,
     withModuleSystem: Boolean
-): Any? {
-    val testModule =
-        when {
-            withModuleSystem ->
-                eval(BasicBoxTest.KOTLIN_TEST_INTERNAL + ".require('" + testModuleName!! + "')")
-            testModuleName === null ->
-                eval("this")
-            else ->
-                get(testModuleName)
-        }
-    testModule as ScriptObjectMirror
-
-    val testPackage =
-        when {
-            testPackageName === null ->
-                testModule
-            testPackageName.contains(".") ->
-                testPackageName.split(".").fold(testModule) { p, part -> p[part] as ScriptObjectMirror }
-            else ->
-                testModule[testPackageName]!!
-        }
+): String? {
+    var script = when {
+        withModuleSystem -> BasicBoxTest.KOTLIN_TEST_INTERNAL + ".require('" + testModuleName!! + "')"
+        testModuleName === null -> "this"
+        else -> testModuleName
+    }
 
-    return (this as Invocable).invokeMethod(testPackage, testFunctionName)
-}
+    if (testPackageName !== null) {
+        script += "[$testPackageName]"
+    }
 
-fun ScriptEngine.loadFile(path: String) {
-    eval("load('${path.replace('\\', '/')}');")
+    val testPackage = eval<Any>(script)
+    return callMethod(testPackage, testFunctionName)
 }
 
-fun ScriptEngine.runAndRestoreContext(
-    globalObject: ScriptObjectMirror = eval("this") as ScriptObjectMirror,
+
+fun InteropEngine.runAndRestoreContext(
+    globalObject: GlobalRuntimeContext = evalAsMap("this"),
     originalState: Map<String, Any?> = globalObject.toMap(),
-    f: ScriptEngine.() -> Any?
+    f: InteropEngine.() -> Any?
 ): Any? {
     return try {
         this.f()
     } finally {
         for (key in globalObject.keys) {
-            globalObject[key] = originalState[key] ?: ScriptRuntime.UNDEFINED
+            if (originalState[key] == null) {
+                globalObject.remove(key)
```

Comment:
If I recall correctly removing globals doesn't really work with Nashorn. At least that was the case at some point and that's why it was being set to undefined instead of being removed.

Also does removing globals work correctly in v8? (Just wondering, it probably does, but that's worth checking =))